crates/masonry/src/text2/edit.rs

```diff
@@ -97,19 +97,55 @@ impl<T: EditableText> TextEditor<T> {
                 if !(mods.control_key() || mods.alt_key() || mods.super_key()) {
                     match &event.logical_key {
                         Key::Named(NamedKey::Backspace) => {
-                            eprintln!("Got backspace, not yet handled");
-                            Handled::No
+                            if let Some(selection) = self.with.selection {
+                                if !selection.is_caret() {
+                                    self.text_mut().edit(selection.range(), "");
+                                    self.with.selection =
+                                        Some(Selection::caret(selection.min(), Affinity::Upstream));
+                                } else if let Some(o) =
+                                    self.text().prev_grapheme_offset(selection.active)
+                                {
+                                    self.text_mut().edit(o..selection.active, "");
+                                    self.with.selection =
+                                        Some(Selection::caret(o, selection.active_affinity));
+                                }
+                                Handled::Yes
+                            } else {
+                                Handled::No
+                            }
+                        }
+                        Key::Named(NamedKey::Delete) => {
+                            if let Some(selection) = self.with.selection {
+                                if !selection.is_caret() {
+                                    self.text_mut().edit(selection.range(), "");
+                                    self.with.selection = Some(Selection::caret(
+                                        selection.min(),
+                                        Affinity::Downstream,
+                                    ));
+                                } else if let Some(o) =
```

Comment:
I don't think `o` is a good name for this variable; we're not rationing characters